### ISSUE TITLE
[1.3.2.1 Windows] Can't import css from npm package (inside another css)

### ISSUE THREAD

**Issue by tcastelli** [meteor/meteor#6846]
I have followed #6037 instructions, installed fourseven:scss and I have loaded scss files from external packages inside my local "atmosphere" package.

The problem comes after that, since when I try to load any external css from another css in the atmosphere package i get 
````
warn: there are some @import rules those are not taking effect as they are required to be in the beginning of the file
````
This happens also when I try to import a CSS from a CSS in my app code, and the only way to solve it, is to import the css from a javascript file. (In my case this is not a solution because of load order problems, so I would need to attacha that CSS to the head styles instead of doing an insert of the style with a require)

To reproduce this, simply install any npm package with scss, then the fourserven:scss with autoprefixer package and try to load CSS from another CSS file with @import "{}/node_modules/...."

**Comment by skirunman** [comment]
From my understanding is you don't need to (and can't) import CSS files from an Atmosphere package. The CSS is automatically evaluated and bundled. You can import Less, Sass and Stylus files as long as they are marked as "import" style files in the package. 

The syntax to import Less, Sass or Stylus files from an Atmosphere package is to use the `{}` package name syntax:

```less
@import '{my-package:pretty-buttons}/buttons/styles.import.less';
```

CSS files in an Atmosphere package are declared with `api.addFiles`, and therefore will be eagerly evaluated, and automatically bundled with all the other CSS in your app.

More info: 

An important feature shared by all of the available CSS pre-processors is the ability to import files. This lets you split your CSS into smaller pieces, and provides a lot of the same benefits that you get from JavaScript modules:

1. You can control the load order of files by encoding dependencies through imports, since the load order of CSS matters.
2. You can create reusable CSS "modules" that just have variables and mixins and don't actually generate any CSS.

In Meteor, each of your `.scss`, `.less`, or `.styl` source files will be one of two types: "source" or "import".

A "source" file is evaluated eagerly and adds its compiled form to the CSS of the app immediately.

An "import" file is evaluated only if imported from some other file and can be used to share common mixins and variables between different CSS files in your app.

Read the documentation for each Pre-processor package to see how to indicate which files are source files vs. imports.

**Comment by tcastelli** [comment]
Is not only from atmosphere packages, but from my app code. And i find strange that we can import css from packages in JS code and not from CSS files (in that case it should be noted in the docs and okay...)

Also imagine this simple scenario:

-  An app in designed with old 1.2 structure where all the code/css are loaded eagerly, so css goes to merged_stylesheets in the head
- This app is using bootstrap/font-awesome, etc packages which include their css in the head in merged_stylesheets (in order to appear before the previous custom css and the order works)

Now imagine we want to switch to the new npm ecosystem and we start migrating from atmosphere packages to npm packages.
Well, in this situation, if we for example want to include bootstrap in our app, since we have the css from the code loaded in the merged_stylesheets , if we do imports from {}/node_modules/bootstrap/..csss from our javascript code, this css will be placed AFTER the merged_stylesheets so everything gets broken.
As a workaround, I created a local atmosphere package to load bootstrap, font-awesome, css from npm packages BEFORE the merged_stylesheets (as you said by exporting with api.useFile). The problem is that when I import external packages css from the npm packages, it will fail and then this is not a good strategy anymore

**Comment by skirunman** [comment]
I agree that it makes sense to be able to lazily load a `.css` file from an npm package and have Meteor's CSS build tool process the file and bundle it with the rest of the apps CSS. I had expected this to work in this way.

```
// import from some .css file in my app
@import 'npm-package-name/stylesheets/styles.css';
```

When importing a `.css` file from JS it makes more sense that the file would be kept separately and just added to the `<head>` tag after the rest of the app's bundled CSS.

@benjamn Is this supposed to work in the way we are discussing or is this a new feature request? Thanks~

**Comment by tcastelli** [comment]
I am creating a small repo so @benjamn can try it out easily if he confirms that imports "should" work from css files.

**Comment by tcastelli** [comment]
Created a repo for easy reproduction https://github.com/tcastelli/csstest

**Comment by skirunman** [comment]
Just curious why you are trying to us import from a Meteor Atmosphere package versus directly from your app written in a Meteor 1.3 style with modules from `imports/` directory? I would not expect to be able to import CSS into a package as I think you have to use [Npm.depends ](http://guide.meteor.com/writing-packages.html#npm-dependencies) to access npm packages from inside a Meteor Atmosphere package.

Can you try import from 'client/local/` first (eager evaluation) and then try it from `imports/local/` (lazy evaluation)?

**Comment by tcastelli** [comment]
-  My app is not following the 1.3 imports strategy yet (just read my scenario in the previous post), this is just s quick repo to demonstrate the problem.
-  Since it is possible to import SCSS files contained in NPM packages from an atmosphere package I guess it should be okay to also import CSS files. Especially knowing that importing from JS is working fine.
- It doesn't work with NPM depends and since that will eventually disappear(that's the point of the new modules system) I wouldn't like a workaround using NPM.depends.
- Not sure If I get your last point correctly, but I have tried to place the css inside packages/local/imports/loadOrder.css and now there is no error, but the css file is not loaded in head.
Also tried /packages/local/loadOrder.css and this time I get the same error, and tries to load the css from http://localhost:3000/bootstrap/dist/css/bootstrap.css

**Comment by tcastelli** [comment]
Ah okay!!! I guess you mean that I should try to load the css outside the package.
I have tried to import directly from my main.css file and it worked!
the bootstrap.css gets loaded, however there's still an error in the console trying to load 
http://localhost:3000/bootstrap/dist/css/bootstrap.css

**Comment by tcastelli** [comment]
Okay I've found something that works, which is kind of dirty and give some errors in the console but...

in client/main.css
````
@import "{}/packages/test_local/loadOrder.css";
````

in packages/local/loadOrder.css
`````
@import "bootstrap/dist/css/bootstrap.css";
````

I still think there has to be a simpler way though

**Comment by lorensr** [comment]
I get the same error when trying `@import` from the top of a CSS file. Only works when importing from JS. But I'd rather the CSS be part of the CSS bundle.

**Comment by skirunman** [comment]
I agree that it would be most desirable to import a `.css` file located inside of an npm package (inside node_modules) using `@import` syntax from another `.css` file in your application (outside of node_modules) so that all the CSS is bundled up and minimized, instead of importing from `.js` file and having the imported `.css` file loaded via the head tag.

**Comment by iameli** [comment]
>Just curious why you are trying to use @import from inside a Meteor Atmosphere package versus directly from your app written in a Meteor 1.3 style with modules from imports/ directory?

I'm currently running into this. We have a number of Meteor applications that all share styles using a package. In Meteor 1.2, that file had an `_includes.scss` that pulled in a lot of NPM modules:

```
// Import font-awesome's variables and mixins so we can include them in ::befores and such
@import ".npm/package/node_modules/font-awesome/scss/variables";
@import ".npm/package/node_modules/font-awesome/scss/mixins";

// Import sass-flexbox-mixin
@import ".npm/package/node_modules/sass-flex-mixin/flex";
```

And that worked great. Now I need to get the same thing using 1.3, and I'm stuck. I can't import these files from a JS file -- they're filled with SCSS `@mixin` statements, they don't actually have any CSS in them, they're only useful if they can be imported by other SCSS files, which they can't.

I'm probably going to work around it for now by just cloning all the relevant node_modules and committing them into the package, but that's not ideal.

**Comment by dnish** [comment]
Any progress on this topic? I want to bundle the imported script with all other CSS code.

**Comment by NitroBAY** [comment]
Seriously... Meteor's team wants Great Leap Forward towards npm but we cannot simply import .css. It's not a complex problem, everything work if the extensions is one of those of preprocessors but just css doesn't work, could you please fix that ?

**Comment by KeitIG** [comment]
Well, you can import css from npm packages with an import in a .js file. But my problem with that is it's not bundled with the rest of the styles. It just goes in a `style` tag somewhere in the DOM.

**Comment by NitroBAY** [comment]
Damn it's the first time a message popup in such a way, I've been surprised !

**Comment by NitroBAY** [comment]
Now I want a very very simple thing : using normalize.css in my app. This npm package does not offer a minified css version, so if I load it via js it's not minified... What am I supposed to do ?

**Comment by rafinskipg** [comment]
Seriously. This issue needs more priority

**Comment by JanMP** [comment]
This also breaks any readymade components on npm that depend on fonts. I would very much like to use the Vue-Markdown component to display content with quite a bit of mathematical formulas. But without the fonts there's no readable math. And that's not something you can fix by copying a font around. I'd have to dig through the component and all its dependencies and change all the references to the different fonts they use. And that's just one component and not an entire ui framework.

So this breaks a whole world of useful things that npm has got to offer. That's a major issue. It's been there since 1.3 and it's still there in 1.6. 

This needs to get fixed ASAP.

**Comment by pilarArr** [comment]
I've run into this problem too. I'm trying to import `css` files from my npm  to my main `app.scss`. 
I'd like to share what I've tried. I'm using `Meteor@1.5.1` right now and `fourseven:scss@4.5.4`

This is what I tried that doesn't work:
```
@import 'bootstrap';
@import '../../../node_modules/bootstrap/dist/css/bootstrap.min.css';
@import '../../../node_modules/bootstrap/dist/css/bootstrap.min';
@import '{}/node_modules/bootstrap/dist/css/bootstrap.min';
```

The work arounds that I'm using right now:
1. Import the css into my `main.js` at start up with this:
```
import 'bootstrap/dist/css/bootstrap.min.css';
import 'bootstrap';
```
  The `bootstrap` css works like this but the glyph icons don't, I have to manually copy the fonts files to the `/public` folder from the `bootstrap` npm for them to load. The same happens with leaflet, the marker icons appear as a broken image.

2. Copy the css file to my `/stylesheets` folder and import them in my `app.scss`.
```
@import './bootstrap.min';
```
  The glyphs and icons still don't work if I do this.

3. Change the file extension in the npm package to `scss`. Then I can import them with this:

```
@import '../../../node_modules/bootstrap/dist/css/bootstrap.min';
```
  The glyphs and icons still don't work if I do this.

4. I suppose that loading them in the `<head>` of the `html` would work as well

**Comment by hwillson** [comment]
Hi everyone - thanks for all of the comments and trouble shooting steps. Importing `.css` files from npm packages, using CSS `@import` syntax, is still an issue with the latest version of Meteor (1.5.2), and can definitely be a pain to work around. The [Importing styles from npm](https://guide.meteor.com/using-npm-packages.html#npm-styles) section of the Guide leads people to believe this functionality is supported and should exist, but we realize it isn't working. I've added the `bug` label here for now and will aim to have more updates here shortly. 

Simple reproduction that demonstrates the issue: 
https://github.com/hwillson/meteor-issue-6846

**Comment by stevenjmarsh** [comment]
I'm not sure if this is a proper way to include bootstrap from an npm package, but it appears to be working for me. I mimicked the steps for adding tachyons, [found here](https://github.com/tableflip/meteor-tachyons),  but npm installed bootstrap instead of tachyons. 
And included...
```
@import "bootstrap/dist/css/bootstrap.css";
``` 
in a `css-packages.css` file in a `./client/stylesheets` folder.
This process uses juliancwirko:postcss. 
I did not see any bootstrap style or file reference within the <head> tags.

**Comment by hwillson** [comment]
Thanks @stevenjmarsh - yes, using PostCSS can help address some of this.

@KeitIG @NitroBAY @rafinskipg and anyone else interested, depending on how the npm package you want to use is setup, you can likely work around this using [`juliancwirko:postcss`](https://atmospherejs.com/juliancwirko/postcss). Using @NitroBAY's request (https://github.com/meteor/meteor/issues/6846#issuecomment-285885109) as an example:

> Now I want a very very simple thing : using normalize.css in my app. This npm package does not offer a minified css version, so if I load it via js it's not minified... What am I supposed to do ?

You could achieve this in your app by:

1) `meteor npm install --save normalize.css`
2) `meteor remove standard-minifier-css`
3) `meteor add juliancwirko:postcss`
4) `meteor npm install --save-dev postcss-easy-import`
5) Add the following to your `package.json`

```
{
  ...
  "postcss": {                                                                       
    "plugins": {                                                                     
      "postcss-easy-import": {}                                                      
    }                                                                                
  }
}
```

6) Add `@import 'normalize.css/normalize.css';` to one of your app's CSS files; e.g. `/client/main.css`
7) Restart your app.

After you app restarts, take a look at the contents of your `merged-stylesheets.css` file. You'll see that the `@import` was actually replaced by the full contents of `node_modules` located `normalize.css` file. This means that when you bundle for production, it will be minified (you can also verify this locally by using `meteor --production`).

This isn't a perfect solution, as it won't address issues like broken font references (e.g. when trying to use this method with [`components-font-awesome`](https://www.npmjs.com/package/components-font-awesome)). Hopefully it helps reduce some frustration until this issue can be addressed further in the Meteor core.

**Comment by KoenCa** [comment]
Do you think this issue will be fixed in Meteor 1.6?

**Comment by hwillson** [comment]
@KoenCa 1.6 is now in its release candidate phase, so its functionality list is pretty much final at this point. Hopefully we can get this addressed in an upcoming (and not too distant) 1.6.x point release.

**Comment by shishirsingh66** [comment]
- Saved 'node_modules/xxx/yyy.css' as 'node_modules/xxx/_yyy.scss'
- in main.scss, added =>  @import '../node_modules/xxx/yyy'
- duplicated any images or fonts folders from node_modules/xxx/ to public/images, public/fonts under meteor tree 
- started meteor; 
- seem to work fine.
- Above as I was using Angular 4 widgets with meteor 1.5.2
Would of course be great if could use simple import statement in .scss, nearest to the component.

**Comment by hwillson** [comment]
Just a quick FYI - `postcss` is now being used by Meteor internally (will be released in Meteor 1.6.1. - see https://github.com/meteor/meteor/pull/9263). This should open up some interesting possibilities with regards to addressing this issue.
